### PR TITLE
Implemented new C++ library for zone administration (main)

### PR DESCRIPTION
**Pull request by korydraughn**
This PR was created from work in #6593 that was mostly complete. This also helps to get these features into upstream faster since there are fewer things to think about.

### CODE REVIEW

**Comment by korydraughn** [comment]
After playing with the clang-format config, it appears we cannot configure clang-format to ignore PP directives.

Seems our only options are:
- Adjust the code so that it isn't weird to look at
- Use `// clang-format off|on` to force clang-format to ignore PP directives
- Simply ignore the results of clang-format github workflow for PP directives

I can also see a big reason to switch from `AfterHash` to `BeforeHash`. That being grepping for macros.

Will have to think about this a bit more.

**Comment by trel** [comment]
Definitely don't like the 'just ignore' option.

**Comment by korydraughn** [comment]
I agree. I'd like to see all workflows green.

I don't see an issue with allowing devs to use either of the first two options. That gives them a choice as to how to best format the PP directives. Reviewers can highlight issues with formatting based on the context.

File: .clang-format [hunk 1]
```diff
@@ -13,8 +13,8 @@ TabWidth:        4
 IndentWidth:     4
 ConstructorInitializerIndentWidth: 4
 
-IndentPPDirectives: AfterHash
-PPIndentWidth:   2
+#IndentPPDirectives: AfterHash
+#PPIndentWidth:   2
```
Comment (korydraughn) [inline .clang-format:17]: This needs to be restored or changed to `BeforeHash`. See https://github.com/irods/irods/pull/6647#issuecomment-1279768353

File: lib/administration/zone/include/irods/zone_administration.hpp [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+            std::string name;        ///< The name of a user or group.
+            // clang-format on
+        }; // struct zone_collection_acl_property
+
+        // clang-format off
+        template <typename T>
+        concept ZoneProperty = std::same_as<T, zone_name_property> ||
+            std::same_as<T, connection_info_property> ||
+            std::same_as<T, comment_property> ||
+            std::same_as<T, zone_collection_acl_property>;
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_administration.hpp:102]: Doxygen does not understand concepts.

Leave a normal comment or see if there's a way to make Doxygen produce documentation for it.

File: lib/administration/zone/include/irods/zone_administration.hpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    // NOLINTBEGIN(cppcoreguidelines-macro-usage)
+    #define NAMESPACE_IMPL client
+    #define RxComm         RcComm
+    #define rxGeneralAdmin rcGeneralAdmin
+    // NOLINTEND(cppcoreguidelines-macro-usage)
+    
+    #include "irods/generalAdmin.h"
+
+struct RcComm;
+#endif // IRODS_ZONE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_administration.hpp:32]: Wrap in clang-format off/on comments.

File: lib/administration/zone/include/irods/zone_administration.hpp (hunk 2 above)
Comment (trel) [reply lib/administration/zone/include/irods/zone_administration.hpp:102]: can bump Doxygen if necessary without too much trouble.

File: .clang-format (hunk 1 above)
Comment (korydraughn) [reply .clang-format]: Reverted changes that attempted to disable PP formatting.

File: lib/administration/zone/include/irods/zone_administration.hpp (hunk 3 above)
Comment (korydraughn) [reply lib/administration/zone/include/irods/zone_administration.hpp:33]: done

File: lib/administration/zone/include/irods/zone_administration.hpp (hunk 2 above)
Comment (korydraughn) [reply lib/administration/zone/include/irods/zone_administration.hpp:111]: Verified that doxygen can handle concepts just fine. No need to bump anything.

File: lib/administration/zone/include/irods/zone_administration.hpp [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <type_traits>
+
+namespace irods::experimental::administration
+{
+    /// \since 4.3.1
+    inline namespace v1
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_administration.hpp:53]: These inline namespaces don't offer much and should be removed.

File: lib/administration/zone/include/irods/zone_administration.hpp [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    /// This namespace's name changes based on the presence of a special macro. If the following macro
+    /// is defined, then the NAMESPACE_IMPL will be \p server, else it will be \p client.
+    ///
+    ///     - IRODS_ZONE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+    ///
+    /// \since 4.3.1
+    namespace NAMESPACE_IMPL
+    {
+        /// \since 4.3.1
+        inline namespace v1
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_administration.hpp:149]: Remove.

File: lib/administration/zone/include/irods/zone_type.hpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
+#ifndef IRODS_ZONE_TYPE_HPP
+#define IRODS_ZONE_TYPE_HPP
+
+/// \file
+
+#include <cstdint>
+
+namespace irods::experimental::administration::inline v1
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_type.hpp:8]: Remove inline namespace

File: lib/administration/zone/src/zone_administration.cpp [hunk 7]
```diff
@@ -0,0 +1,100 @@
+#include "irods/zone_administration.hpp"
+
+#ifdef IRODS_ZONE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+// The server maintains a list of the zones known to the system.
+// The library takes advantage of this information where possible to avoid
+// unnecessary network and database operations.
+extern zoneInfo* ZoneInfoHead; // NOLINT(cppcoreguidelines-avoid-non-const-global-variables)
+#endif // IRODS_ZONE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+
+namespace irods::experimental::administration::NAMESPACE_IMPL::inline v1
```
Comment (korydraughn) [inline lib/administration/zone/src/zone_administration.cpp:10]: Remove inline namespace

File: unit_tests/src/test_zone_administration.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    REQUIRE(zone_info.has_value());
+    CHECK(zone_info->type == adm::zone_type::remote);
+    CHECK(zone_info->name == test_zone_name_1);
+    CHECK(zone_info->connection_info == conn_info_2);
+    CHECK(zone_info->comment == comment_2);
+
+    // Show that the handling of the zone collection in regards to strict-acls can be modified.
+    const adm::zone_collection_acl_property prop{adm::zone_collection_acl::read, "rods"};
+    CHECK_NOTHROW(adm::client::modify_zone(conn, test_zone_name_1, prop));
+
```
Comment (alanking) [inline unit_tests/src/test_zone_administration.cpp:117]: Only suggestion I have is to maybe add a test or two that exercises the error cases, i.e. `CHECK_THROWS`.

**Comment by alanking** [comment]
Seems good to me. The inline namespace stuff is interesting to me, not familiar with that. Also, is the introduction of the feature test macro setting a precedent that any new feature needs to add a new macro?

File: .clang-format (hunk 1 above)
Comment (korydraughn) [reply .clang-format]: Leaving `AfterHash` as the default. Developers can choose how they want to format PP directives using clang-format off/on.

Reviewers will will the approval regarding formatting.

File: unit_tests/src/test_zone_administration.cpp (hunk 8 above)
Comment (korydraughn) [reply unit_tests/src/test_zone_administration.cpp:117]: I've added some basic tests for error cases.

**Comment by korydraughn** [comment]
All unit tests pass.

I don't think the core tests need to be run as this only introduces a wrapper library and swizzles some cmake.

If there aren't any objectives, I'll add the #'s and we can get this merged.

File: lib/administration/zone/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+foreach(variant IN ITEMS client server)
+  set(target_name "irods_zone_administration_${variant}")
+
+  add_library(
+    ${target_name}
+    OBJECT
+    "${CMAKE_CURRENT_SOURCE_DIR}/src/zone_administration.cpp"
+  )
+  # TODO Is this needed?
```
Comment (korydraughn) [inline lib/administration/zone/CMakeLists.txt:12]: Remove this.

**Review by alanking — APPROVED** [review]
Seems fine to me.

I'll poke @SwooshyCueb just in case he has any thoughts on the feature test bits or anything else in this PR :)

File: lib/administration/zone/include/irods/zone_administration.hpp [hunk 10]
```diff
@@ -0,0 +1,288 @@
+#ifndef IRODS_ZONE_ADMINISTRATION_HPP
+#define IRODS_ZONE_ADMINISTRATION_HPP
+
+/// \file
+
+#undef NAMESPACE_IMPL
+#undef RxComm
+#undef rxGeneralAdmin
+
+// clang-format off
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_administration.hpp:10]: remove

File: lib/administration/zone/include/irods/zone_administration.hpp [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+#  define NAMESPACE_IMPL client
+#  define RxComm         RcComm
+#  define rxGeneralAdmin rcGeneralAdmin
+// NOLINTEND(cppcoreguidelines-macro-usage)
+
+#  include "irods/generalAdmin.h"
+
+struct RcComm;
+#endif // IRODS_ZONE_ADMINISTRATION_ENABLE_SERVER_SIDE_API
+// clang-format on
```
Comment (korydraughn) [inline lib/administration/zone/include/irods/zone_administration.hpp:34]: remove

File: lib/administration/zone/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+  )
+  target_include_directories(
+    ${target_name}
+    PUBLIC
+    "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    PRIVATE
+    "$<BUILD_INTERFACE:${CMAKE_IRODS_BINARY_DIR}/lib/core/include>"
+    "$<BUILD_INTERFACE:${CMAKE_IRODS_SOURCE_DIR}/lib/core/include>"
+    "$<BUILD_INTERFACE:${CMAKE_IRODS_SOURCE_DIR}/lib/api/include>"
```
Comment (SwooshyCueb) [inline lib/administration/zone/CMakeLists.txt:25]: `$<BUILD_INTERFACE` include dirs should be `PUBLIC`

**Review by SwooshyCueb — CHANGES_REQUESTED** [review]

File: lib/administration/zone/CMakeLists.txt (hunk 12 above)
Comment (korydraughn) [reply lib/administration/zone/CMakeLists.txt:24]: Done.

**Comment by korydraughn** [comment]
Resolved all comments.

Any remaining thoughts before I add the #'s?

**Comment by alanking** [comment]
None from me

**Comment by trel** [comment]
clang format has a couple opinions, empty lines...

otherwise, issue numbers check out... good for me.

**Comment by korydraughn** [comment]
Clang-format complaining about empty lines is incorrect. It shouldn't remove those. Even github knows to complain about source files missing the trailing empty line.

**Comment by korydraughn** [comment]
Added #'s

File: lib/administration/zone/CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+  )
+  target_include_directories(
+    ${target_name}
+    PUBLIC
+    "$<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>"
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>"
+    "$<BUILD_INTERFACE:${CMAKE_IRODS_BINARY_DIR}/lib/core/include>"
+    "$<BUILD_INTERFACE:${CMAKE_IRODS_SOURCE_DIR}/lib/core/include>"
+    "$<BUILD_INTERFACE:${CMAKE_IRODS_SOURCE_DIR}/lib/api/include>"
+    "${IRODS_EXTERNALS_FULLPATH_BOOST}/include"
```
Comment (SwooshyCueb) [inline lib/administration/zone/CMakeLists.txt:25]: you still want the `IRODS_EXTERNALS_FULLPATH` includes to be private.

**Comment by SwooshyCueb** [comment]
oh oops, didn't realize it had already been merged. I'll submit a PR for it in the morning

**Pull request merged**